src/cluster.py

```diff
@@ -721,32 +793,59 @@ def remove_raft_member(self, member_ip: str) -> None:
             RaftMemberNotFoundError: if the member to be removed
                 is not part of the raft cluster.
         """
+        if self.charm.has_raft_keys():
+            logger.debug("Remove raft member: Raft already in recovery")
+            return
+
         # Get the status of the raft cluster.
-        raft_status = subprocess.check_output([
-            "charmed-postgresql.syncobj-admin",
-            "-conn",
-            "127.0.0.1:2222",
-            "-pass",
-            self.raft_password,
-            "-status",
-        ]).decode("UTF-8")
+        syncobj_util = TcpUtility(password=self.raft_password, timeout=3)
+
+        raft_host = "127.0.0.1:2222"
+        try:
+            raft_status = syncobj_util.executeCommand(raft_host, ["status"])
+        except UtilityException:
+            logger.warning("Remove raft member: Cannot connect to raft cluster")
+            raise RemoveRaftMemberFailedError()
 
         # Check whether the member is still part of the raft cluster.
-        if not member_ip or member_ip not in raft_status:
+        if not member_ip or f"partner_node_status_server_{member_ip}:2222" not in raft_status:
+            return
+
+        # If there's no quorum and the leader left raft cluster is stuck
+        if not raft_status["has_quorum"] and (
+            not raft_status["leader"] or raft_status["leader"].host == member_ip
+        ):
+            logger.warning("Remove raft member: Stuck raft cluster detected")
+            data_flags = {"raft_stuck": "True"}
+            try:
+                health_status = self.get_patroni_health()
+            except Exception:
+                logger.warning("Remove raft member: Unable to get health status")
+                health_status = {}
+            if health_status.get("role") in ("leader", "master") or health_status.get(
+                "sync_standby"
+            ):
+                logger.info("%s is raft candidate" % self.charm.unit.name)
+                data_flags["raft_candidate"] = "True"
+            self.charm.unit_peer_data.update(data_flags)
```

Comment:
Initial flags that trigger the recovery attempt. If we need to disable this (stable releases?), it would be easiest from here.